### PR TITLE
Fix Shell Navigation for Hierarchally registered Global Routes

### LINKED ISSUE #11237: [Bug] Xamarin.Forms Shell - contextual navigation from global routes is not working
### Description
Having two pages registered as global routes (see "Register page routes" under [Shell navigation](https://docs.microsoft.com/en-us/xamarin/xamarin-forms/app-fundamentals/shell/navigation#routes)) and navigating in a contextual way won't work (nothing happens, no exception, no return value or anything else).

### Steps to Reproduce

1. Open the provided example -> [Xaminals global routes contextual navigation.zip](https://github.com/xamarin/Xamarin.Forms/files/5802296/Xaminals.global.routes.contextual.navigation.zip)
2. Open it in your IDE.
3. Open the file Xaminals.ViewModels.MonkeyDetailViewModel.cs.
4. Add a breakpoint at line 49.
5. Start the app and navigate to 
5.1 Monkeys (tab)
5.2 Select any monkey
5.3 Click the button "Navigate to MonkeyGenomePage"

### Expected Behavior

App should navigate to the page MonkeyGenomePage.

### Actual Behavior

Nothing happens.

### Basic Information

- Version with issue: Xamarin.Forms 4.7.0.968
- Last known good version: -
- IDE: Visual Studio for Mac 8.5
- Platform Target Frameworks: 
  - iOS:  Simulator 13.5
  - Android: 9.0
- Android Support Library Version: 28.0.03

### Workaround

Not using contextual routes on global routes + using a prefix to make routes unique.

### LINKED ISSUE #13328: [Bug] [Shell] [XF5.0.0] Unable to navigate using PopAsync on Shell
### Description
Unable to navigate back using ```Shell.Current.Navigation.PopAsync()``` on third child Shell page ```MainPage/FirstPage/SecondPage/ThirdPage```. I'm not sure if this is a bug or if it's designed to work like that.

### Steps to Reproduce

1. Start the application
2. Click on "Click Me!" buttons 3 times until you get a page with a label displaying "Third Page"
3. Use the mobile hardware back button to go back 

### Expected Behavior
* User is navigated back to the "Second Page"

### Actual Behavior
* User is unable to go back

### Basic Information

- Version with issue: Xamarin.Forms 5.0.0.1874
- Last known good version: Xamarin.Forms 4.8.0 (Latest)
- Platform Target Frameworks: 
  - Android: 11 
- NuGet Packages: Xamarin.Essentials 1.6.0
- Affected Devices: Samsung Galaxy S9+ SM-G965F (Android 10)

### Environment

<details><summary>Show/Hide Visual Studio info</summary>

```
Visual Studio: Visual Studio Professional 2019 16.8.3
```

</details>

### Reproduction Link

https://github.com/elicul/XamarinPopAsyncBug/tree/master

### Workaround

* Use ```Shell.Current.Navigation.RemovePage(Shell.Current.Navigation.NavigationStack.LastOrDefault());``` to remove the last page

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change ###

Fixed Hierarchical Global Route Registration so that it would correctly apply the route registered to the current state of shell

If a user is registering full route path  `Route.Register("path1/path2/path3")` and Shell is currently at `"path1/path2"`

Then the user should be able to navigate to that route via
- `GotoAsync("path3")` or `Route.Register("path1/path2/path3")`

### Issues Resolved ### 
- fixes #13328
- fixes #11237

### Platforms Affected ### 
- Core/XAML (all platforms)

### Testing Procedure ###
- unit test included

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: Xamarin.Forms.Core/Shell/NavigationRequest.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+﻿using System.Diagnostics;
+
+namespace Xamarin.Forms
+{
+	[DebuggerDisplay("RequestDefinition = {Request}, StackRequest = {StackRequest}")]
+	internal class NavigationRequest
```
Comment (PureWeen) [inline Xamarin.Forms.Core/Shell/NavigationRequest.cs:6]: No changes here just moving this class to its own file

File: Xamarin.Forms.Core/Shell/RequestDefinition.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Xamarin.Forms
+{
+	[DebuggerDisplay("Full = {FullUri}, Short = {ShortUri}")]
+	internal class RequestDefinition
+	{
+		public RequestDefinition(RouteRequestBuilder theWinningRoute, Shell shell)
```
Comment (PureWeen) [inline Xamarin.Forms.Core/Shell/RequestDefinition.cs:10]: No changes here just moving this class to its own file

File: Xamarin.Forms.Core/Shell/RouteRequestBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,217 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Xamarin.Forms
+{
+	/// <summary>
+	/// This attempts to locate the intended route trying to be navigated to
+	/// </summary>
+	internal class RouteRequestBuilder
```
Comment (PureWeen) [inline Xamarin.Forms.Core/Shell/RouteRequestBuilder.cs:10]: No changes here just moving this class to its own file

**Comment by PureWeen** [comment]
Failing Tests are unrelated

**Review by hartez — APPROVED** [review]

File: Xamarin.Forms.Core/Shell/ShellNavigationManager.cs [hunk 4]
```diff
@@ -490,46 +490,31 @@ public static ShellNavigationState GetNavigationState(ShellItem shellItem, Shell
...
+			return BuildFlattenedNavigationStack(section.Stack, section.Navigation.ModalStack);
+		}
+		
+		public static List<Page> BuildFlattenedNavigationStack(IReadOnlyList<Page> startingList, IReadOnlyList<Page> modalStack)
+		{
+			var returnValue = startingList.ToList();
+			if (modalStack == null)
+				return returnValue;
 
-			List<string> CollapsePath(
```
Comment (PureWeen) [inline Xamarin.Forms.Core/Shell/ShellNavigationManager.cs:495]: This good just got moved to `ShellUriHandler.CollapsePath`

File: Xamarin.Forms.Core/Shell/ShellSection.cs [hunk 5]
```diff
@@ -475,24 +475,6 @@ void RemoveExcessPathsWithinTheRoute()
 			}
 		}
 
-		List<Page> BuildFlattenedNavigationStack(List<Page> startingList, IReadOnlyList<Page> modalStack)
```
Comment (PureWeen) [inline Xamarin.Forms.Core/Shell/ShellSection.cs:478]: This code just got moved to `ShellNavigationManager.BuildFlattenedNavigationStack`

**Review by rmarinho — APPROVED** [review]

**Pull request merged**